tools/tmt/README.md

```diff
@@ -0,0 +1,283 @@
+will not print the map contents to the console.  This is because even
+small applications can contain thousands of entries and the whole map
+content is better written to a file instead of producing voluminous
+output on the screen.
+
+## Filtering
+
+### Type name filtering
+
+Passing a regular expression to the utility causes it to match type
+names (both Managed and Java for **each** loaded entry, regardless on
+whether the current map being processed is java-to-managed or
+managed-to-java) against it and report the matches to the console.
+
+Also, in this case the files are not generated by the default, which
+can be changed by using the `-g` option.  This is because filtering
+mode is mostly useful for "quick look" at the maps and the full output
+will likely be of no interest in this instance.
+
+### Architecture filtering
+
+By default the utility reads `libxamarin-app.so` for any architecture
+it finds in the location indicated on command line.  The `-a` option
+can be used to limit the set of architectures by specifying a
+comma-separated list of architecture names from the following set
+(architecture names are case-insensitive): `ARM`, `ARM64`, `X86` and
+`X86_64`
+
+# Command line options
+
+```
+Usage: tmt [OPTIONS] <FILE.apk|FILE.aab|libxamarin-app.so|PROJECT_DIR> [FILTER_
+REGEX]
+
+OPTIONS are:
+
+  -j, --only-java            Process only the java-to-managed map
+  -m, --only-managed         Process only the managed-to-java map
+  -s, --short-report         Omit some map details from the report (e.g. MVID
+                               and TokenID from managed-to-java map)
+  -o, --output-directory=DIR Write the report files in the DIR directory
+                               instead of the current one
+  -a, --arch=ARCH_LIST       Limit reporting only to the specified
+                               architectures. ARCH_LIST is a comma-separated
+                               list of architectures (one of, case-insensitive:
+                               ARM, ARM64, X86, X86_64)
+  -1, --only-first           Process only the first shared library from the APK/
+                               AAB archive or a project directory. Architecture
+                               filter is ignored in this case. This is the
+                               default action when regex filtering is used,
+                               otherwise it defaults to `false`
+  -g, --generate-files       Generate report files. If regex filtering is used,
+                               this setting defaults to `false`, it is `true`
+                               otherwise.
+
+  -v, --verbose              Show debug messages
+```
+
+## Sample invocations
+
+### Dump all typemaps in a project directory
+``` shell
+# 
+# Alternative invocations with the same result:
+#
+# bin/Debug/bin/tmt/tmt samples/HelloWorld/obj
+# bin/Debug/bin/tmt/tmt samples/HelloWorld/obj/Debug
+#
+$ bin/Debug/bin/tmt/tmt samples/HelloWorld/
```

Comment:
I built this branch, and the `bin/Debug/bin/tmt/tmt` script does not exist.

In order to run the app, I need to use `dotnet bin/Debug/bin/tmt/tmt.dll`, which is fine and works (as far as it goes), but doesn't match this doc.

Additionally, `tools/tmt/run.sh` doesn't look like it's supposed to be copied into `bin/Debug/bin/tmt/tmt`, as that contains an `msbuild /r` invocation…

Is something missing from this PR which is still on your local tree?

Reply:
`tmt` is a netcore app and it produces `bin/Debug/tmt/tmt` as a native executable, it's not a script:

```shell
$ ls -l bin/Debug/bin/tmt/tmt
-rwxr-xr-x 1 grendel grendel 86584 Feb  3 22:20 bin/Debug/bin/tmt/tmt

$ file bin/Debug/bin/tmt/tmt
bin/Debug/bin/tmt/tmt: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=22531e6f94df2fbec39dd3d1ef003bcdcc0288df, stripped
```

Not sure why you don't have it in your tree.

`run.sh` will be removed, it's merely a shortcut I use for testing to avoid typing the same commands all over again.